File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-doc-5-4/ContractIdentifier.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class ContractIdentifier : ISerialized
    {
        public ContractIdentifier(XmlNode xmlNode)
        {
            XmlNode partyReferenceNode = xmlNode.SelectSingleNode("partyReference");
            
            if (partyReferenceNode != null)
            {
                if (partyReferenceNode.Attributes["href"] != null || partyReferenceNode.Attributes["id"] != null) 
                {
                    if (partyReferenceNode.Attributes["id"] != null) 
                    {
                        partyReferenceIDRef_ = partyReferenceNode.Attributes["id"].Value;
                        PartyReference ob = new PartyReference(partyReferenceNode);
                        IDManager.SetID(partyReferenceIDRef_, ob);
                    }
                    else if (partyReferenceNode.Attributes["href"] != null)
                    {
                        partyReferenceIDRef_ = partyReferenceNode.Attributes["href"].Value;
                    }
                    else
                    {
                        partyReference_ = new PartyReference(partyReferenceNode);
                    }
                }
                else
                {
                    partyReference_ = new PartyReference(partyReferenceNode);
                }
            }
            
        
            XmlNodeList contractIdNodeList = xmlNode.SelectNodes("contractId");
            
            if (contractIdNodeList != null)
            {
                this.contractId_ = new List<ContractId>();
                foreach (XmlNode item in contractIdNodeList)
                {
                    if (item.Attributes["href"] != null || item.Attributes["id"] != null) 
                    {
                        if (item.Attributes["id"] != null) 
                        {
                            contractIdIDRef_ = item.Attributes["id"].Value;
                            contractId_.Add(new ContractId(item));
                            IDManager.SetID(contractIdIDRef_, contractId_[contractId_.Count - 1 ]);
                        }
                        else if (item.Attributes["href"] != null)
                        {
                            contractIdIDRef_ = item.Attributes["href"].Value;
                        }
                        else
                        {
                        contractId_.Add(new ContractId(item));
                        }
                    }
                    else
                    {
                        contractId_.Add(new ContractId(item));
                    }
                }
            }
            
        
            XmlNodeList versionedContractIdNodeList = xmlNode.SelectNodes("versionedContractId");
            
            if (versionedContractIdNodeList != null)
            {
                this.versionedContractId_ = new List<VersionedContractId>();
                foreach (XmlNode item in versionedContractIdNodeList)
                {
                    if (item.Attributes["href"] != null || item.Attributes["id"] != null) 
                    {
                        if (item.Attributes["id"] != null) 
                        {
                            versionedContractIdIDRef_ = item.Attributes["id"].Value;
                            versionedContractId_.Add(new VersionedContractId(item));
                            IDManager.SetID(versionedContractIdIDRef_, versionedContractId_[versionedContractId_.Count - 1 ]);
                        }
                        else if (item.Attributes["href"] != null)
                        {
                            versionedContractIdIDRef_ = item.Attributes["href"].Value;
                        }
                        else
                        {
                        versionedContractId_.Add(new VersionedContractId(item));
                        }
                    }
                    else
                    {
                        versionedContractId_.Add(new VersionedContractId(item));
                    }
                }
            }
            
        
        }
        
    
        #region PartyReference_
        private PartyReference partyReference_;
        public PartyReference PartyReference_
        {
            get
            {
                if (this.partyReference_ != null)
                {
                    return this.partyReference_; 
                }
                else if (this.partyReferenceIDRef_ != null)
                {
                    partyReference_ = IDManager.getID(partyReferenceIDRef_) as PartyReference;
                    return this.partyReference_; 
                }
                else
                {
                      return this.partyReference_; 
                }
            }
            set
            {
                if (this.partyReference_ != value)
                {
                    this.partyReference_ = value;
                }
            }
        }
        #endregion
        
        public string partyReferenceIDRef_ { get; set; }
        #region ContractId_
        private List<ContractId> contractId_;
        public List<ContractId> ContractId_
        {
            get
            {
                if (this.contractId_ != null)
                {
                    return this.contractId_; 
                }
                else if (this.contractIdIDRef_ != null)
                {
                    return this.contractId_; 
                }
                else
                {
                      return this.contractId_; 
                }
            }
            set
            {
                if (this.contractId_ != value)
                {
                    this.contractId_ = value;
                }
            }
        }
        #endregion
        
        public string contractIdIDRef_ { get; set; }
        #region VersionedContractId_
        private List<VersionedContractId> versionedContractId_;
        public List<VersionedContractId> VersionedContractId_
        {
            get
            {
                if (this.versionedContractId_ != null)
                {
                    return this.versionedContractId_; 
                }
                else if (this.versionedContractIdIDRef_ != null)
                {
                    return this.versionedContractId_; 
                }
                else
                {
                      return this.versionedContractId_; 
                }
            }
            set
            {
                if (this.versionedContractId_ != value)
                {
                    this.versionedContractId_ = value;
                }
            }
        }
        #endregion
        
        public string versionedContractIdIDRef_ { get; set; }
        public string choiceStr_0;
        
    
        
    
    }
    
}

File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-doc-5-4/ContractIdentifier.cpp
// ContractIdentifier.cpp 
#include "ContractIdentifier.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

ContractIdentifier::ContractIdentifier(TiXmlNode* xmlNode)
: ISerialized(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //partyReferenceNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* partyReferenceNode = xmlNode->FirstChildElement("partyReference");

   if(partyReferenceNode){partyReferenceIsNull_ = false;}
   else{partyReferenceIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- partyReferenceNode , address : " << partyReferenceNode << std::endl;
   #endif
   if(partyReferenceNode)
   {
      if(partyReferenceNode->Attribute("href") || partyReferenceNode->Attribute("id"))
      {
          if(partyReferenceNode->Attribute("id"))
          {
             partyReferenceIDRef_ = partyReferenceNode->Attribute("id");
             partyReference_ = boost::shared_ptr<PartyReference>(new PartyReference(partyReferenceNode));
             partyReference_->setID(partyReferenceIDRef_);
             IDManager::instance().setID(partyReferenceIDRef_,partyReference_);
          }
          else if(partyReferenceNode->Attribute("href")) { partyReferenceIDRef_ = partyReferenceNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { partyReference_ = boost::shared_ptr<PartyReference>(new PartyReference(partyReferenceNode));}
   }

   //contractIdNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* contractIdNode = xmlNode->FirstChildElement("contractId");

   if(contractIdNode){contractIdIsNull_ = false;}
   else{contractIdIsNull_ = true;}

   if(contractIdNode)
   {
      for(contractIdNode; contractIdNode; contractIdNode = contractIdNode->NextSiblingElement("contractId")){
          #ifdef ConsolePrint
              FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- contractIdNode , address : " << contractIdNode << std::endl;
          #endif
          if(contractIdNode->Attribute("href") || contractIdNode->Attribute("id"))
          {
              if(contractIdNode->Attribute("id"))
              {
                  contractIdIDRef_ = contractIdNode->Attribute("id");
                  contractId_.push_back(boost::shared_ptr<ContractId>(new ContractId(contractIdNode)));
                  contractId_.back()->setID(contractIdIDRef_);
                  IDManager::instance().setID(contractIdIDRef_, contractId_.back());
              }
              else if(contractIdNode->Attribute("href")) { contractIdIDRef_ = contractIdNode->Attribute("href");}
              else { QL_FAIL("id or href error"); }
          }
          else { contractId_.push_back(boost::shared_ptr<ContractId>(new ContractId(contractIdNode)));}
      }
   }
   else {
       #ifdef ConsolePrint
           FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- contractIdNode , address : " << contractIdNode << std::endl;
       #endif
   }

   //versionedContractIdNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* versionedContractIdNode = xmlNode->FirstChildElement("versionedContractId");

   if(versionedContractIdNode){versionedContractIdIsNull_ = false;}
   else{versionedContractIdIsNull_ = true;}

   if(versionedContractIdNode)
   {
      for(versionedContractIdNode; versionedContractIdNode; versionedContractIdNode = versionedContractIdNode->NextSiblingElement("versionedContractId")){
          #ifdef ConsolePrint
              FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- versionedContractIdNode , address : " << versionedContractIdNode << std::endl;
          #endif
          if(versionedContractIdNode->Attribute("href") || versionedContractIdNode->Attribute("id"))
          {
              if(versionedContractIdNode->Attribute("id"))
              {
                  versionedContractIdIDRef_ = versionedContractIdNode->Attribute("id");
                  versionedContractId_.push_back(boost::shared_ptr<VersionedContractId>(new VersionedContractId(versionedContractIdNode)));
                  versionedContractId_.back()->setID(versionedContractIdIDRef_);
                  IDManager::instance().setID(versionedContractIdIDRef_, versionedContractId_.back());
              }
              else if(versionedContractIdNode->Attribute("href")) { versionedContractIdIDRef_ = versionedContractIdNode->Attribute("href");}
              else { QL_FAIL("id or href error"); }
          }
          else { versionedContractId_.push_back(boost::shared_ptr<VersionedContractId>(new VersionedContractId(versionedContractIdNode)));}
      }
   }
   else {
       #ifdef ConsolePrint
           FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- versionedContractIdNode , address : " << versionedContractIdNode << std::endl;
       #endif
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<PartyReference> ContractIdentifier::getPartyReference()
{
   if(!this->partyReferenceIsNull_){
        if(partyReferenceIDRef_ != ""){
             return boost::shared_static_cast<PartyReference>(IDManager::instance().getID(partyReferenceIDRef_));
        }else{
             return this->partyReference_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<PartyReference>();
   }
}
std::vector<boost::shared_ptr<ContractId>> ContractIdentifier::getContractId()
{
   if(!this->contractIdIsNull_){
        if(contractIdIDRef_ != ""){
             return this->contractId_;
        }else{
             return this->contractId_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return std::vector<boost::shared_ptr<ContractId>>();
   }
}
std::vector<boost::shared_ptr<VersionedContractId>> ContractIdentifier::getVersionedContractId()
{
   if(!this->versionedContractIdIsNull_){
        if(versionedContractIdIDRef_ != ""){
             return this->versionedContractId_;
        }else{
             return this->versionedContractId_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return std::vector<boost::shared_ptr<VersionedContractId>>();
   }
}
}

